### DISCUSSION TITLE
what's it used for the modified rentention params of pulsar in milvus

### DISCUSSION THREAD

**Discussion by wangqia0309** [milvus-io/milvus#34755] [Q&A and General discussion]
in milvus k8s mode deploy conf, defaultRetention related params of pulsar is custom configured，not the default value 0 that stands for no retention data. i wonder why did it？ if the action can help recovery from accidental offline to querynode or datanode, can you please explain the reason in depth.  According to my own opinion, neither datanode nor querynode would send ack to pulsar before they finished the data practically, so the data in pulsar should not be cleaned even if the datanode or querynode is down temporarily

**Comment by xiaofan-luan** [comment]
> in milvus k8s mode deploy conf, defaultRetention related params of pulsar is custom configured，not the default value 0 that stands for no retention data. i wonder why did it？ if the action can help recovery from accidental offline to querynode or datanode, can you please explain the reason in depth. According to my own opinion, neither datanode nor querynode would send ack to pulsar before they finished the data practically, so the data in pulsar should not be cleaned even if the datanode or querynode is down temporarily

You r right, milvus has to ack only on data is consumed and put to S3.
But current implementation is milvus ack every time after consume(Obviously this is a wrong implementation but a history problem). So we need to reserve data in pulsar for at least 1 day in case milvus fails and data need to be recovered from mq.
Usually 3 day is more recommended

**Reply by wangqia0309** [reply]
did you only answer the behavior of datanode? would querynode which is for realtime indexing recover from the stop point in pulsar after it restart from once downtime(supposed the segment is not closed)? or to say, is it able to restore indexing from the previous stop point without retention data?

**Reply by xiaofan-luan** [reply]
both querynode and datanode share the same checkpoint.
if data is saved to S3 already, then querynode will load from S3 as well
